### DISCUSSION TITLE
✨ Parity with `docker run`

### DISCUSSION THREAD

**Discussion by helderco** [dagger/dagger#12260] [Ideas]
This is related to:
- https://github.com/dagger/dagger/issues/8140

## Summary

Create a convenience in the API for a `Container.run` function that executes the entrypoint + default arguments and returns the combined `stdout` and `stderr` output.

## Motivation

It was brought to my attention by @jpadams that, because the CLI is more and more powerful, he wanted to play with the `alpine/git` image but expected it to work more like `docker run`. It’s simple to see what the image does by default with docker:

```console
❯  docker container run -it --rm alpine/git

usage: git [-v | --version] [-h | --help] [-C <path>] [-c <name>=<value>]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           [--config-env=<name>=<envvar>] <command> [<args>]
```

As you can see, it has an entrypoint and default args:

```console
❯ dagger core container from --address=alpine/git

_type: Container
defaultArgs:
    - --help
entrypoint:
    - git
mounts: []
platform: linux/arm64
user: ""
workdir: /git
```

But in dagger, it’s much more convoluted to do the same in the CLI (or even in code for that matter). While before you’d just add `stdout` to the command, now you need this:

```console
❯ dagger core container from --address=alpine/git with-exec --args= --use-entrypoint stdout
```

But what if we added a convenience `Container.run`, that executes the entrypoint and default arguments, not the last `withExec`, and returns (ideally) the stream of the output?

```console
❯ dagger container from --address=alpine/git run

usage: git [-v | --version] [-h | --help] [-C <path>] [-c <name>=<value>]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | -P | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           [--config-env=<name>=<envvar>] <command> [<args>]
```

*Note that I used the second option from https://github.com/dagger/dagger/issues/6947#issuecomment-2168278578 (`dagger <core function>`). See https://github.com/dagger/dagger/pull/8135.*

Basically making this:
```go
out, err := ctr.Run(ctx)
```
A convenience for:
```go
// See https://github.com/dagger/dagger/issues/3496 for combined output idea
out, err := ctr.WithExec([]string{}, dagger.ContainerWithExecOpts{UseEntrypoint: true}).Output(ctx)
```

**Comment by M-Pixel** [comment]
Dagger as an alternative to Docker/Podman was quite useful for me, in particular for getting the contents of certain directories of distroless containers that don't contain `ls`, and for quickly testing the behavior of certain Dagger commands.

The ability to specify default arguments without naming them would go a long way.  Here are a few ideas to consider:

```
dagger dag container=alpine/git run
```

`dag` instead of `core`.  Aligns with the way that other SDKs work (`core` makes the CLI tantilizingly close to being an SDK itself).  `=` on an object as shorthand for positional parameters.  `run`, as in your suggestion, a shorthand for `with-exec --args= --use-entrypoint stdout`.

What about objects/functions with multiple parameters?  Treat them as a heterogeneous array.  What if one of the parameters is an array? [ JSON array syntax support would fix that](https://github.com/dagger/dagger/issues/9426).

```
dagger dag container=alpine/git with-directory=.,. run=status

# Maybe even treat run specially, where subsequent parameters are forwarded
dagger dag container=alpine/git with-directory=.,. run status

# While explicit names are great for writing a pipeline as TypeScript or Go, and although some people will abuse this and
# actually write shell scripts this way, for truly interactive CLI use, short aliases of various commands would help
dagger dag ctr=alpine/git +dir=.,. run status

# compare that to
podman run --rm -v .:/mnt -w /mnt alpine/git status
```
